File: code/bonus_chapters/wordcount/python/wordcount_by_reducebykey_shorthand.py
from __future__ import print_function
import sys
from pyspark.sql import SparkSession
#======================================
#
# NOTE: print() and collect() are used for debugging and educational purposes.
#
# @author Mahmoud Parsian
#
#======================================
def main():

    # create an instance of a SparkSession as spark
    spark = SparkSession.builder.getOrCreate()

    # set input path
    input_path = sys.argv[1]
    print("input_path=", input_path)

    frequencies = spark.sparkContext.textFile(input_path)\
        .flatMap(lambda line: line.split(" "))
        .map(lambda word: (word, 1))\
        .reduceByKey(lambda a, b: a + b)
    #
    print(frequencies.collect())

    # done!
    spark.stop()
#end-def
#======================================
if __name__ == "__main__":
    main()

File: code/bonus_chapters/wordcount/scala/src/main/scala/org/data/algorithms/spark/bonuschapter/WordCountByReduceByKeyShorthand.scala
package org.data.algorithms.spark.bonuschapter

import org.apache.spark.sql.SparkSession

/**
 *----------------------------------------
 * NOTE: print() and collect() are used for debugging and educational purposes.
 *
 * @author Biman Mandal
 *----------------------------------------
 */
object WordCountByReduceByKeyShorthand {
  def main(args: Array[String]): Unit = {
    // create an instance of a SparkSession as spark
    val spark = SparkSession.builder.master("local[*]").getOrCreate()

    // set input path
    val inputPath = args(0)
    println("inputPath=" + inputPath)

    val frequencies = spark.sparkContext.textFile(inputPath)
      .flatMap(line => line.split(" "))
      .map(word => (word, 1))
      .reduceByKey((x, y) => x + y)

    println(frequencies.collect().mkString("Array(", ", ", ")"))

    // done!
    spark.stop()
  }

}
